### PR TITLE
fix getUnion returning duplicate items from both sets

### LINKED ISSUE #83: getUnion() returns duplicate items
If the union of sets A and B is neither A nor B nor null, the lists of items _a_ in A and _b_ in B are filtered and then concatenated by `getUnion(A, B, a, b, algebra)`.  However, these lists are not further filtered for uniqueness by the id clause (nor by equality by all properties, nor even by object sameness).

This is shown in https://jsbin.com/megitokize/edit?html,js,output -- the union of items should only contain one object with _id = 2, but two are returned, one from being in `items1` and one from being in `items2`.

This leads to incorrect lists returned from connections with the cache-requests behavior; items may be displayed twice as the union of the cached set and the fetched set is stored as the new cached set.

### PR DESCRIPTION
**Pull request by bmomberger-bitovi**
Fixes #83

### CODE REVIEW

File: src/get_test.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ var set = require("./set");
 var QUnit = require("steal-qunit");
 var props = require("./props");
 var h = require("./helpers");
+var assign = require("can-util/js/assign/assign");
```
Comment (justinbmeyer) [inline src/get_test.js]: can you use `can-assign`?

Comment (bmomberger-bitovi) [reply src/get_test.js]: Certainly can.

File: src/set-core.js [hunk 2]
```diff
@@ -6,6 +6,33 @@ var assign = require("can-util/js/assign/assign");
 var each = require("can-util/js/each/each");
 var makeArray = require("can-util/js/make-array/make-array");
 var isEmptyObject = require("can-util/js/is-empty-object/is-empty-object");
+var getProps = require("can-util/js/get/get");
+
+// concatUnique
+// concat all items in bItems onto aItems that do not already exist in aItems.
+// same-object and ID collisions are both looked at when deciding whether
+// an item matches another.
```
Comment (justinbmeyer) [inline src/set-core.js:14]: why both same-object and ID?  Wouldn't ID cover both cases?

File: src/set-core.js [hunk 3]
```diff
@@ -6,6 +6,33 @@ var assign = require("can-util/js/assign/assign");
...
+function concatUnique(aItems, bItems, algebra) {
+	var keyTree = {};
```
Comment (justinbmeyer) [inline src/set-core.js]: might want to call this `idTree` ... only to distinguish with `can-key-tree`.

File: src/set-core.js [hunk 4]
```diff
@@ -2,10 +2,37 @@ var h = require("./helpers");
 var clause = require("./clause");
 var compare = require("./compare");
 var get = require("./get");
-var assign = require("can-util/js/assign/assign");
+var assign = require("can-assign");
 var each = require("can-util/js/each/each");
 var makeArray = require("can-util/js/make-array/make-array");
 var isEmptyObject = require("can-util/js/is-empty-object/is-empty-object");
+var getProps = require("can-util/js/get/get");
```
Comment (justinbmeyer) [inline src/set-core.js]: I would alias this as `getPropValue` or `getProp` as it only returns a single property value.

File: src/set-core.js [hunk 5]
```diff
@@ -2,10 +2,37 @@ var h = require("./helpers");
...
+	aItems.forEach(function(item) {
+		var keyNode = keyTree;
+		each(algebra.clauses.id, function(prop) {
+			keyNode = keyNode[getProps(item, prop)] = keyNode[getProps(item, prop)] || {};
+		});
+	});
+
+	return aItems.concat(bItems.filter(function(item) {
+		var keyNode = keyTree;
+		if(aItems.indexOf(item) > -1) {
```
Comment (justinbmeyer) [inline src/set-core.js]: this makes this concat `O(n^2)`.  Use of `Set` would be nice here ... or drop this b/c `id` "should" take care of it.

File: src/set-core.js (hunk 2 above)
Comment (bmomberger-bitovi) [reply src/set-core.js:14]: Same object covers the case where there is no ID clause.  Ideally I'd like to cover same-object and equal object by all props when ID is absent, but that would be slower and too much work for the time being.

File: src/set-core.js (hunk 5 above)
Comment (bmomberger-bitovi) [reply src/set-core.js]: Set (CIDSet for compatibility) is a good idea.  As mentioned above, if there is no `id` clause, testing for same object will catch duplicates in some use cases.

File: src/set-core.js (hunk 4 above)
Comment (bmomberger-bitovi) [reply src/set-core.js]: now getProp

**Review by justinbmeyer — COMMENTED** [review]
nice work!

**Pull request merged**